pkg/istio-agent/sds-agent.go

```diff
@@ -320,76 +326,43 @@ func ingressSdsExists() bool {
-	ret := &secretfetcher.SecretFetcher{}
+	fetcher := &secretfetcher.SecretFetcher{}
 
 	// TODO: get the MC public keys from pilot.
 	// In node agent, a controller is used getting 'istio-security.istio-system' config map
 	// Single caTLSRootCert inside.
 
 	var err error
 
-	// TODO: this should all be packaged in a plugin, possibly with optional compilation.
-	log.Infof("serverOptions.CAEndpoint == %v", serverOptions.CAEndpoint)
-	if (serverOptions.CAProviderName == "GoogleCA" || strings.Contains(serverOptions.CAEndpoint, "googleapis.com")) &&
-		stsclient.GKEClusterURL != "" {
-		// Use a plugin to an external CA - this has direct support for the K8S JWT token
-		// This is only used if the proper env variables are injected - otherwise the existing Citadel or Istiod will be
-		// used.
-		caClient, err = gca.NewGoogleCAClient(serverOptions.CAEndpoint, true)
-		serverOptions.PluginNames = []string{"GoogleTokenExchange"}
-	} else {
-		// Determine the default CA.
-		// If /etc/certs exists - it means Citadel is used (possibly in a mode to only provision the root-cert, not keys)
-		// Otherwise: default to istiod
-		//
-		// If an explicit CA is configured, assume it is mounting /etc/certs
-		var rootCert []byte
-
-		tls := true
-		certReadErr := false
-
-		if serverOptions.CAEndpoint == "" {
-			// When serverOptions.CAEndpoint is nil, the default CA endpoint
-			// will be a hardcoded default value (e.g., the namespace will be hardcoded
-			// as istio-system).
-			log.Info("Istio Agent uses default istiod CA")
-			serverOptions.CAEndpoint = "istiod.istio-system.svc:15012"
-
-			if serverOptions.PilotCertProvider == "istiod" {
-				log.Info("istiod uses self-issued certificate")
-				if rootCert, err = ioutil.ReadFile(path.Join(CitadelCACertPath, constants.CACertNamespaceConfigMapDataName)); err != nil {
-					certReadErr = true
-				} else {
-					log.Infof("the CA cert of istiod is: %v", string(rootCert))
-				}
-			} else if serverOptions.PilotCertProvider == "kubernetes" {
-				log.Infof("istiod uses the k8s root certificate %v", k8sCAPath)
-				if rootCert, err = ioutil.ReadFile(k8sCAPath); err != nil {
-					certReadErr = true
-				}
-			} else if serverOptions.PilotCertProvider == "custom" {
-				log.Infof("istiod uses a custom root certificate mounted in a well known location %v",
-					cache.DefaultRootCertFilePath)
-				if rootCert, err = ioutil.ReadFile(cache.DefaultRootCertFilePath); err != nil {
-					certReadErr = true
-				}
-			} else {
-				certReadErr = true
-			}
-			if certReadErr {
-				rootCert = nil
-				// for debugging only
-				log.Warnf("Failed to load root cert, assume IP secure network: %v", err)
-				serverOptions.CAEndpoint = "istiod.istio-system.svc:15010"
-				tls = false
-			}
+	// If proxy is using file mounted certs, we do not have to connect to CA.
+	if shouldProvisionCertificates() {
```

Comment:
There is no logic change here. Just moved every thing in to this if block and moved some CA specific settings of `workloadSdsCacheOptions` inside if

Reply:
Can we instead do
```
if !shouldProvisionCertificates() {
   ...
   return
}
```
to avoid the big if block?

Reply:
Done